Add examples & docs

thgs opened the pull request:
If somebody has an idea for another example that they would like to see here, please let me know and I will add.

Any comments are more than welcome. I saw this part being empty of usage documentation so I gave it a try to fill it in. However because docs can be very subjective on how one wants them to be, please comment on it freely. 

If this part is not in the right or preferred direction (and is not easily fixed) maybe it makes sense to just keep the example.

thgs committed:
Add some documentation and an example

kelunik commented on README.md:
````diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+```php
+$session->lock();
+$session->set('key', $data);
+$session->commit();     // commits & unlocks
````
```suggestion
$session->commit(); // commits & unlocks
```

thgs replied:
changed

kelunik commented on README.md:
````diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+To read data from the session is straightforward:
+
+```php
+$session->get('key');       // will read data stored in key 'key'
````
```suggestion
$session->get('key'); // will read data stored in key 'key'
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+
+Note that `get()` will return `null` if the data in `key` are not found.
+
+In order to write data, the session must be `lock()`ed first so that it cannot be written from anywhere else
```
```suggestion
In order to write data, the session must be `lock()`ed first so that it cannot be written from anywhere else.
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+
+Calling `commit()` will store the data in the session storage and unlock the session.
+
+Other important methods that belong in the Session are:
```
```suggestion
Other important methods of the `Session` class are:
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+
+#### Use the middleware to access Session in a RequestHandler
+
+As this package is a plugin for [amphp/http-server](https://github.com/amphp/http-server) there is a middleware
```
```suggestion
As this package is a plugin for [`amphp/http-server`](https://github.com/amphp/http-server) there is a middleware
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+The middleware will handle setting and reading a session cookie in the request/response as well as releasing all locks
+on the session after the request has been processed.
+
+For instructions on how to use the middlewares with `amphp/http-server` see [amphp/http-server Middleware section](https://github.com/amphp/http-server#middleware).
```
```suggestion
If you haven't used middleware in `amphp/http-server`, follow the [instructions on how to use middleware with `amphp/http-server`](https://github.com/amphp/http-server#middleware).
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+
+For instructions on how to use the middlewares with `amphp/http-server` see [amphp/http-server Middleware section](https://github.com/amphp/http-server#middleware).
+
+A simple example is provided here [examples/simple](https://github.com/amphp/http-server-session/blob/3.x/examples/simple.php).
```
```suggestion
A simple example is provided here [`examples/simple.php`](https://github.com/amphp/http-server-session/blob/3.x/examples/simple.php).
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+
+Internally the session works with 3 dependencies:
+
+* [KeyedMutex](https://github.com/amphp/sync/blob/2.x/src/KeyedMutex.php) - A synchronisation primitive to be used across contexts
```
```suggestion
* [`KeyedMutex`](https://github.com/amphp/sync/blob/2.x/src/KeyedMutex.php) - A synchronisation primitive to be used across contexts
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+Internally the session works with 3 dependencies:
+
+* [KeyedMutex](https://github.com/amphp/sync/blob/2.x/src/KeyedMutex.php) - A synchronisation primitive to be used across contexts
+* [SessionStorage](https://github.com/amphp/http-server-session/blob/3.x/src/SessionStorage.php) - Interface for reading and writing data.
```
```suggestion
* [`SessionStorage`](https://github.com/amphp/http-server-session/blob/3.x/src/SessionStorage.php) - Interface for reading and writing data.
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+
+* [KeyedMutex](https://github.com/amphp/sync/blob/2.x/src/KeyedMutex.php) - A synchronisation primitive to be used across contexts
+* [SessionStorage](https://github.com/amphp/http-server-session/blob/3.x/src/SessionStorage.php) - Interface for reading and writing data.
+* [SessionIdGenerator](https://github.com/amphp/http-server-session/blob/3.x/src/SessionIdGenerator.php) - Interface for generating and validating
```
```suggestion
* [`SessionIdGenerator`](https://github.com/amphp/http-server-session/blob/3.x/src/SessionIdGenerator.php) - Interface for generating and validating
```

thgs replied:
Changed

kelunik commented on README.md:
```diff
@@ -12,7 +12,107 @@ composer require amphp/http-server-session
+  a session ID.
+
+An instance of the [Session](https://github.com/amphp/http-server-session/blob/3.x/src/Session.php#L28) can be constructed easily using
+the provided [SessionFactory](https://github.com/amphp/http-server-session/blob/3.x/src/SessionFactory.php)
```
```suggestion
A [`Session`](https://github.com/amphp/http-server-session/blob/3.x/src/Session.php#L28) instance can be constructed using
the provided [`SessionFactory`](https://github.com/amphp/http-server-session/blob/3.x/src/SessionFactory.php).
```

thgs replied:
Changed

kelunik commented on examples/simple.php:
```diff
@@ -0,0 +1,71 @@
+$server->start($middlewareStack, $errorHandler);
+
+// Await a termination signal to be received.
+$signal = trapSignal([\SIGHUP, \SIGINT, \SIGQUIT, \SIGTERM]);
```
```suggestion
$signal = trapSignal([\SIGINT, \SIGTERM]);
```

thgs replied:
Changed

thgs committed:
Add documentation for SessionMiddleware configuration

thgs commented:
@kelunik changes done, added section for CookieAttributes and also added some code on the example, 

however, I think there is something wrong with how the example works. I had tested it and was progressing properly (+1) but now it seems it does (+2). So let me see into that.

thgs commented:
@kelunik fixed that issue, changes done. 

It was calls to `favicon.ico`. Oddly chrome calls everytime twice while firefox calls once at the first request 🤷 . Anyhow, now the example is progressing +1 every reload.
